### ISSUE TITLE
Precision loss in large-scale FCNMV operators (tolerance relaxed to 5e-2)

### ISSUE THREAD

**Issue by Hepbmstl** [chaobrain/brainevent#108]
- [x] Provide a complete example of how to reproduce the bug, wrapped in triple backticks like this:
- [x] Check for duplicate issues.
- [x] If applicable, include full error messages/tracebacks.

### Description
In large-scale matrix tests (`m=4000`, `conn=4000`), the new custom operators (`compact_binary_fcnmv`, `bitpack_binary_fcnmv`, `binary_fcnmv`) fail to pass the default precision check (`rtol=1e-3, atol=1e-3`) when compared to the dense reference implementation. 

**Key Finding:** The precision loss is highly dependent on weight heterogeneity.
* When `homo_w=True`, max absolute difference is `0.0`.
* When `homo_w=False` (heterogeneous weights), max absolute difference up to `~3.5e-2`. 

This issue may be caused by the precision problem of ref due to the use of floating-point numbers in event dispatch operations by jax.

*Note: The tolerance in the CI has been temporarily relaxed to `5e-2` as a workaround to unblock the pipeline.*

### Reproduction

```bash
# Compact Binary
pytest compact_binary_test.py::test_compact_binary_fcnmv_forward_in_large_scale

# Bitpack Binary
pytest bitpack_binary_test.py::test_bitpack_fcnmv_forward_in_large_scale

# more details: 
[4.8.md](https://github.com/user-attachments/files/26558556/4.8.md)

**Comment by Hepbmstl** [comment]
Environment
OS: Linux

Hardware: NVIDIA GeForce RTX 3090 (24GB)

Python Stack: Python 3.12.13, pytest 9.0.2

CUDA Stack: Driver 590.48.01 / CUDA 13.1 / nvidia-cuda-nvcc 13.2.51

JAX Stack: jax 0.8.3, jaxlib 0.8.3, jax-cuda13-plugin 0.8.3

Target Package: brainevent 0.0.7
